packages/providers/src/fuel-graphql-request.ts

```diff
@@ -0,0 +1,154 @@
+  `fetch` implementation, reducing the final bundle size by 15.68 kB.
+
+  More:
+    https://github.com/FuelLabs/fuels-ts/discussions/1592#discussioncomment-8003515
+
+*/
+
+import { FuelError } from '@fuel-ts/errors';
+import type { DocumentNode } from 'graphql';
+import { print } from 'graphql';
+
+function isSubscription({ definitions: defs }: DocumentNode) {
+  const opDef = defs.find((op) => op.kind === 'OperationDefinition');
+
+  if (opDef && 'operation' in opDef) {
+    return opDef?.operation === 'subscription';
+  }
+
+  return false;
+}
+
+class FuelGraphqlSubscriber implements AsyncIterator<unknown> {
+  private stream!: ReadableStreamDefaultReader<Uint8Array>;
+  private static textDecoder = new TextDecoder();
+
+  private static parseBytesStream(
+    bytes: Uint8Array | undefined
+  ): Record<string, unknown> | FuelError | undefined {
+    if (bytes === undefined) {
+      return undefined;
+    }
+
+    const text = this.textDecoder.decode(bytes);
+
+    // https://github.com/FuelLabs/fuel-core/blob/e1e631902f762081d2124d9c457ddfe13ac366dc/crates/fuel-core/src/graphql_api/service.rs#L247
+    if (text === 'keep-alive-text') {
+      return undefined;
+    }
+
+    const { data, errors } = JSON.parse(text.split('data:')[1]);
+
+    if (Array.isArray(errors)) {
+      return new FuelError(
+        FuelError.CODES.INVALID_REQUEST,
+        errors.map((err) => err.message).join('\n\n')
+      );
+    }
+
+    return data as Record<string, unknown>;
+  }
+
+  public constructor(
+    private options: {
+      url: string;
+      request: RequestInit;
+      fetchFn: typeof fetch;
+    }
+  ) {}
+
+  private async setStream() {
+    const { url, request, fetchFn } = this.options;
+    const response = await fetchFn(`${url}-sub`, {
+      ...request,
+      headers: {
+        ...request.headers,
+        Accept: 'text/event-stream',
+      },
+    });
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.stream = response.body!.getReader();
+  }
+
+  private async readStream(): Promise<IteratorResult<unknown, unknown>> {
+    const { value, done } = await this.stream.read();
+
+    const parsed = FuelGraphqlSubscriber.parseBytesStream(value);
+
+    if (parsed === undefined && !done) {
+      // we recursively wait for the next message until it's a proper gql response or the stream is done i.e. {value: undefined, done: true}
+      return this.readStream();
+    }
+
+    return { value: parsed, done };
+  }
+
+  async next(): Promise<IteratorResult<unknown, unknown>> {
+    if (!this.stream) {
+      await this.setStream();
+    }
+
+    const { value, done } = await this.readStream();
+    if (value instanceof FuelError) {
+      throw value;
+    }
+
+    return { value, done };
+  }
+
+  /**
+   * Gets called when `break` is called in a `for-await-of` loop.
+   */
+  async return(): Promise<IteratorResult<unknown, undefined>> {
+    await this.stream.cancel();
```

Comment:
Just like `this.stream.read`, `this.stream.cancel()` and `this.stream.releaseLock()` can fail for unexpected reasons. 

Will be safer to add a `Try/Catch` block here and properly handle any potential error.

Reply:
Those errors aren't part of our domain and we'd have to bubble them up to the consumer in the `catch` block anyways, so it's best to leave them as they are and not obfuscate them with custom error handling. It's better for us to to get an error report from an user in this case where the error possibly contains environment-related information than if we got an error message we ourselves created and perhaps obfuscated some crucial environment-related info.

Reply:
@Torres-ssf and I had a discussion where we questioned does calling `ReadableStream.cancel()` actually cancel the underlying connection made with `fetch`? The answer can be found in the example shown on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/cancel#examples). In short, the connection is indeed cancelled.